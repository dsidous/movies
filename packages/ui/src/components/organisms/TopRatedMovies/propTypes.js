import { array } from 'prop-types';

export const propTypes = {
  popular: array.isRequired,
};

export const defaultProps = {};
