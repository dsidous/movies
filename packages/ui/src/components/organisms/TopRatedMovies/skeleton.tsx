import React from 'react';
import Skeleton from '@material-ui/lab/Skeleton';

import useStyles from './styles';

export default () => {
  const classes = useStyles();
  return (
    <section>
      <Skeleton className={classes.header} />
      <div className={classes.root}>
        <Skeleton
          className={`${classes.item} featured`}
          variant="rect"
          height={410}
        />
        <Skeleton className={classes.item} variant="rect" height={200} />
        <Skeleton className={classes.item} variant="rect" height={200} />
        <Skeleton className={classes.item} variant="rect" height={200} />
        <Skeleton className={classes.item} variant="rect" height={200} />
      </div>
    </section>
  );
};
