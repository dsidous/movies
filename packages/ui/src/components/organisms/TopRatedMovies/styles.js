import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      display: 'grid',
      gridGap: '10px',
      gridTemplateColumns: '1fr',
      margin: '0 auto',

      [theme.breakpoints.up('sm')]: {
        gridTemplateColumns: '1.555fr 2fr 2fr',
      },
    },

    header: {
      margin: '40px auto',
    },

    item: {
      cursor: 'pointer',
      position: 'relative',

      '& img': {
        display: 'block',
        transition: 'all 0.3s ease-in',
        width: '100%',
      },

      '&:hover img': {
        opacity: '.8',
        transform: 'scale(.98)',
      },

      '&.featured': {
        [theme.breakpoints.up('sm')]: {
          gridRow: '1 / span 2',
        },
      },
    },

    title: {
      bottom: 10,
      color: '#fff',
      left: 10,
      position: 'absolute',
      textShadow: '0 0 5px #000',
    },

    menu: {
      listStyle: 'none',
      margin: '0 auto 20px',
      padding: 0,

      '& li': {
        cursor: 'pointer',
        display: 'inline',
        margin: '0 10px 0 0',

        '&.active': {
          fontWeight: 'bold',
        },

        '&:focus, & span:focus': {
          outline: 'none',
        },
      },
    },
  }),
  { name: 'featured' },
);
