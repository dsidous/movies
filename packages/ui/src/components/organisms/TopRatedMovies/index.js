import React, { useState } from 'react';
import { Link } from 'react-router-dom';

import { propTypes } from './propTypes';
import MediaImage from '../../atoms/MediaImage';

const topGenres = [
  { id: -1, name: 'All' },
  { id: 28, name: 'Action' },
  { id: 35, name: 'Comedy' },
  { id: 18, name: 'Drama' },
  { id: 10751, name: 'Family' },
  { id: 878, name: 'Science Fiction' },
  { id: 53, name: 'Thriller' },
];

const TopRatedMovies = ({ popular }) => {
  const [active, setActive] = useState(-1);
  const [topMovies, setTopMovies] = useState(popular.slice(0, 5));

  const randomList = arrayMax => {
    const nbrArray = Array.from(Array(arrayMax), (_, x) => x);
    const rndList = [];
    // eslint-disable-next-line no-plusplus
    for (let n = 1; n <= 5; ++n) {
      const rndNum = Math.floor(Math.random() * (arrayMax - n) + 1);
      rndList.push(nbrArray[rndNum]);
      nbrArray[rndNum] = nbrArray[arrayMax - n];
    }
    return rndList;
  };

  const filterTopMovies = genre => {
    let movies =
      genre !== -1
        ? popular.filter(movie => movie.genre_ids.includes(genre))
        : popular;

    const moviesLength = movies.length;

    if (moviesLength > 5) {
      const rndList = randomList(moviesLength);
      movies = movies.filter((movie, index) => rndList.indexOf(index) !== -1);
    }

    setTopMovies(() => movies);
  };

  const selectGenre = topGenreId => {
    setActive(topGenreId);
    filterTopMovies(topGenreId);
  };

  const genresList = () =>
    topGenres.map(topGenre => (
      <li
        data-test={topGenre.id}
        key={topGenre.id}
        className={topGenre.id === active ? 'active' : ''}
      >
        <span
          onClick={() => selectGenre(topGenre.id)}
          onKeyDown={() => selectGenre(topGenre.id)}
          role="button"
          tabIndex="-1"
        >
          {topGenre.name}
        </span>
      </li>
    ));

  return (
    <div>
      {topMovies && (
        <div>
          <h3 className="top-list-main-title">Movies of the day</h3>
          <ul className="top-genres-list">{genresList()}</ul>
          <div className="top-list">
            {topMovies.map((movie, i) => (
              <Link
                to={`/movie/${movie.id}`}
                key={movie.id}
                className={[
                  'top-list__element',
                  i === 0 ? 'featured' : '',
                ].join(' ')}
              >
                {movie.poster_path !== null && i === 0 && (
                  <MediaImage
                    mediaType="poster"
                    size={5}
                    filePath={movie.poster_path}
                    name={movie.title}
                  />
                )}
                {movie.backdrop_path !== null && i !== 0 && (
                  <MediaImage
                    mediaType="poster"
                    size={5}
                    filePath={movie.backdrop_path}
                    name={movie.title}
                  />
                )}
                <div className="top-list__element-title">{movie.title}</div>
              </Link>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default TopRatedMovies;

TopRatedMovies.propTypes = propTypes;
