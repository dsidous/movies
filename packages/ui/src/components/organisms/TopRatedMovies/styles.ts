import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      display: 'grid',
      gridGap: theme.spacing(2),
      gridTemplateColumns: '1fr',
      margin: '0 auto',

      [theme.breakpoints.up('sm')]: {
        gridTemplateColumns: '1.555fr 2fr 2fr',
      },
    },

    header: {
      margin: theme.spacing(5, 'auto'),
    },

    item: {
      cursor: 'pointer',
      position: 'relative',

      '& img': {
        display: 'block',
        transition: 'all 0.3s ease-in',
        width: '100%',
      },

      '&:hover img': {
        opacity: '.8',
        transform: 'scale(.98)',
      },

      '&.featured': {
        [theme.breakpoints.up('sm')]: {
          gridRow: '1 / span 2',
        },
      },
    },

    title: {
      bottom: theme.spacing(2),
      color: '#fff',
      left: theme.spacing(2),
      position: 'absolute',
      textShadow: '0 0 5px #000',
    },

    menu: {
      listStyle: 'none',
      margin: theme.spacing(0, 'auto', 3),
      padding: 0,

      '& li': {
        cursor: 'pointer',
        display: 'inline',
        margin: theme.spacing(0, 2, 0, 0),

        '&.active': {
          fontWeight: theme.fontWeightBold,
        },

        '&:focus, & span:focus': {
          outline: 'none',
        },
      },
    },
  }),
  { name: 'featured' },
);
