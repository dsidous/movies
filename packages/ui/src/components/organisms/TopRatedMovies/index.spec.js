import React, { useState as useStateMock } from 'react';
import { shallow } from 'enzyme';
import TopRatedMovies from '.';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

const mockProps = {
  popular: [
    {
      id: 1,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
    {
      id: 2,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
    {
      id: 3,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
    {
      id: 4,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
    {
      id: 5,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
    {
      id: 6,
      poster_path: 'path/to/poster',
      title: 'title',
      vote_average: '1',
      genre_ids: [28],
    },
  ],
};

describe('Organisms/TopRatedMovies', () => {
  let wrapper;
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation(init => [init, setState]);
    wrapper = shallow(<TopRatedMovies {...mockProps} />);
  });

  it('should render as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should filter movies by genre', () => {
    wrapper.find('[data-test=28] span').props().onClick();

    expect(setState).toHaveBeenCalledWith(28);
  });

  it('should handle filter function for more than 5 result', () => {
    wrapper.find('[data-test=28] span').props().onClick();

    expect(wrapper.find('.top-list__element-title').length).toEqual(5);
  });
});
