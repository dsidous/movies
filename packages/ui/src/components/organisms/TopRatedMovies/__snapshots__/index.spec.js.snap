// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/TopRatedMovies should render as expected 1`] = `ShallowWrapper {}`;
